lib/server/sigv4.js

```diff
@@ -0,0 +1,966 @@
+ * credentials. Handles complete signature verification workflow
+ * including credential lookup, signature calculation, and validation.
+ *
+ * @param {Object} opts Verification options object containing:
+ * @param {Object} opts.req HTTP request object with headers/query
+ * @param {Object} opts.log Bunyan logger instance
+ * @param {Object} opts.redis Redis client for credential lookup
+ * @param {Object} opts.ufds UFDS client for temporary credential lookup
+ * @param {function} cb Callback function
+ * @param {Error} cb.err Error if verification failed
+ * @param {Object} cb.result Verification result containing:
+ *   - user: User object from credential store
+ *   - accessKeyId: Verified access key identifier
+ *   - valid: Boolean verification status
+ *   - isTemporaryCredential: Boolean for temp credential type
+ *   - assumedRole: Role ARN for temporary credentials
+ *
+ * @note Supports both permanent access keys and STS temporary credentials
+ * @note Validates signature timestamp against 15-minute window
+ * @note Handles session token verification for temporary credentials
+ * @note Reconstructs canonical request for signature verification
+ *
+ * @error InvalidSignature Authorization header missing/malformed
+ * @error InvalidSignature Access key not found or invalid
+ * @error InvalidSignature Signature mismatch or expired
+ *
+ * @see AWS Signature Version 4 specification
+ * @since 1.0.0
+ */
+/*jsl:ignore*/
+function verifySigV4(opts, cb) {
+        assert.object(opts, 'opts');
+        assert.object(opts.req, 'opts.req');
+        assert.object(opts.log, 'opts.log');
+        assert.object(opts.redis, 'opts.redis');
+        assert.func(cb, 'callback');
+
+        var req = opts.req;
+        var log = opts.log;
+        var redis = opts.redis;
+        var ufds = opts.ufds; // UFDS client for temporary credential lookup
+
+        log.debug('sigv4.verify: entered');
+
+        var authHeader = req.headers.authorization;
+        if (!authHeader) {
+                setImmediate(cb, new errors.InvalidSignatureError(
+                        'Missing Authorization header'));
+                return;
+        }
+
+        var authInfo = parseAuthHeader(authHeader);
+        if (!authInfo) {
+                setImmediate(cb, new errors.InvalidSignatureError(
+                        'Invalid Authorization header format'));
+                return;
+        }
+
+        // Debug: Log the parsed authorization info
+        log.debug({
+                authHeader: authHeader,
+                parsedAuthInfo: authInfo,
+                accessKeyId: authInfo.accessKeyId,
+                accessKeyIdLength: authInfo.accessKeyId ?
+                        authInfo.accessKeyId.length : 0,
+                accessKeyIdHex: (authInfo.accessKeyId &&
+                                 typeof (authInfo.accessKeyId) === 'string') ?
+                     (function () {
+                        if (authInfo.accessKeyId.length > 1024) {
```

Comment:
If this check is important enough to `throw` I think it may be worthwhile to move it out of this IIFE and up a few lines to have have it return the error with the callback e.g. ` new errors.InvalidSignatureError('Access key ID too long')` as this is the only place we `throw` in this module and clients are likely expecting a callback with an error.

Reply:
@cneira still not sure we should `throw` here.